Enable providers to use other providers from application libctx

mattcaswell opened the pull request:
This is an attempt to fix the issues discussed in #14925.

A provider may want to access algorithms from other providers. Therefore some providers may just want to use the same libctx as used by the application - and therefore reuse all the providers that have already been loaded there.

However a provider may be linked against a different libcrypto to the application (e.g. if one is using static linking and another is using dynamic linking). Therefore it is not possible for the provider to use exactly the *same* libctx as the application.

To get around this we introduce the concept of a "child" libctx. The provider creates a child libctx which is linked to the application's libctx. All the provider's loaded into the application's libctx are automatically mirrored in the child libctx. This works by copying the provctx and OSSL_DISPATCH table used by the "parent" providers in the "parent" libctx into the child providers in the child libctx.

As far as the provider is concerned it appears as if it is using the "same" libctx. Providers are not reloaded and can be seemly reused across the parent and child libctxs.

mattcaswell commented:
This is now essentially working except:

1) The test I added is currently failing due to #15032 

2) I have not attempted to synchronise the global properties between the parent and child libctx. I think this should be relatively straight forward.

3) Needs documentation updates

Before getting this PR into a reviewable state, #14925 needs review by OTC in light of this PR, to check we are going in the right direction.

mattcaswell commented:
Lots of updates to this to properly handle changes in the loaded providers, add tests and missing documentation. This is now essentially finished except it depends on and includes the commits from #15042. I will keep this as WIP until then but is otherwise reviewable.

This does *not* include the ability to track changes in the application's global properties. I will handle that in a separate follow-on PR. This PR is already big enough.

paulidale commented on crypto/provider_core.c:
```diff
@@ -91,8 +100,13 @@ struct ossl_provider_st {
 
+    /* Whether this provider is the child of some other provider */
+    const OSSL_CORE_HANDLE *handle;
+    unsigned int ischild:1;
```
Might this be better conditioned out under FIPS_MODULE.

paulidale commented on crypto/provider_core.c:
```diff
@@ -676,44 +744,134 @@ static int provider_init(OSSL_PROVIDER *prov, int flag_lock)
+    if (!CRYPTO_THREAD_read_lock(store->lock))
         return 0;
+    if (!CRYPTO_THREAD_write_lock(prov->flag_lock))
+        return -1;
```
Release the `store->lock` before returning.

paulidale commented on crypto/provider_core.c:
```diff
@@ -676,44 +744,134 @@ static int provider_init(OSSL_PROVIDER *prov, int flag_lock)
+
+        max = sk_OSSL_PROVIDER_CHILD_CB_num(store->child_cbs);
+        if (lock && !CRYPTO_THREAD_write_lock(prov->flag_lock))
+            return 0;
```
Release `store->lock` before returning.

paulidale commented on crypto/provider_core.c:
```diff
@@ -676,44 +744,134 @@ static int provider_init(OSSL_PROVIDER *prov, int flag_lock)
+    if ((store = get_provider_store(prov->libctx)) == NULL)
+        return 0;
+
+    if (!store->freeing)
```
You might need a rebase.  This needs to be done under a read lock to avoid different bit fields fighting over the memory location.

paulidale commented on crypto/provider_core.c:
```diff
@@ -1022,6 +1197,33 @@ void ossl_provider_unquery_operation(const OSSL_PROVIDER *prov,
+    int i, num, res = 1;
+
+    if ((store = get_provider_store(libctx)) != NULL) {
+        if (!CRYPTO_THREAD_write_lock(store->lock))
```
Definitely need a rebase, this was changed to a read lock.

paulidale commented on crypto/provider_core.c:
```diff
@@ -1068,6 +1270,140 @@ int ossl_provider_test_operation_bit(OSSL_PROVIDER *provider, size_t bitnum,
+
+    if (!CRYPTO_THREAD_write_lock(store->lock))
+        return;
+    max = sk_OSSL_PROVIDER_CHILD_CB_num(store->child_cbs);
```
What about using `sk_OSSL_PROVIDER_CHILD_CB_find` ?

mattcaswell replied:
`sk_OSSL_PROVIDER_CHILD_CB_find` helps you find a specific entry in the stack. Here we need to loop over all of them.

paulidale commented on test/p_test.c:
```diff
@@ -109,6 +115,45 @@ static int p_get_params(void *provctx, OSSL_PARAM params[])
+             * just fallen back to default.
+             */
+#ifdef PROVIDER_INIT_FUNCTION_NAME
+            EVP_MD *md4 = EVP_MD_fetch(ctx->libctx, "MD4", NULL);
```
This will need conditioning on `OPENSSL_NO_MD4`

paulidale commented on test/p_test.c:
```diff
@@ -109,6 +115,45 @@ static int p_get_params(void *provctx, OSSL_PARAM params[])
+             * ourselves, and the legacy provider which we inherit from the
+             * parent libctx. We should also have "this" provider available.
+             */
+            if (OSSL_PROVIDER_available(ctx->libctx, "default")
```
Is diagnostic output possible here to distinguish the failure cases?

mattcaswell replied:
I'm not sure how to emit diagnostic output from inside this test provider?

paulidale replied:
What about raising an error?

mattcaswell replied:
I've raised different error codes for different types of failure cases.

paulidale commented on include/internal/cryptlib.h:
```diff
@@ -164,9 +165,15 @@ typedef struct ossl_ex_data_global_st {
+# define OSSL_LIB_CTX_METHOD_PRIORITY_2             2
 
 typedef struct ossl_lib_ctx_method {
+    int priority;
```
I generally prefer float for priorities (they allow insertion between existing values) but since this is internal and we control everything, I don't see a need.

paulidale reviewed (changes requested):
This looks quite good.

mattcaswell commented:
All feedback addressed or responded to above. I've rebased on the latest version of #15042 and will rebase again once that is merged.

I also fixed a nasty bug relating to built-in providers.

mattcaswell commented:
The final outstanding comment has been resolved and I've rebased to set the latest fips checksums to avoid conflicts with master.

paulidale commented on crypto/ex_data.c:
```diff
@@ -331,6 +334,27 @@ int CRYPTO_dup_ex_data(int class_index, CRYPTO_EX_DATA *to,
+        return 1;
+    if (bp->excb == NULL)
+        return -1;
+    if (ap->excb->priority == bp->excb->priority)
```
What about `return bp->excb->priority - ap->excb->priority;` instead of this and the subsequent two lines?

paulidale commented on crypto/provider_child.c:
```diff
@@ -0,0 +1,331 @@
+
+    /*
+     * We need an OSSL_LIB_CTX but c_get_libctx returns OPENSSL_CORE_CTX. We are
+     * a built-in provider and so we can get away with this cast. Normal
```
This will come back to bite us ....

paulidale commented on crypto/provider_child.c:
```diff
@@ -0,0 +1,331 @@
+    gbl->curr_prov = prov;
+
+    if ((cprov = ossl_provider_find(ctx, provname, 1)) != NULL) {
+
```
Spurious blank line.

openssl-machine commented:
This pull request is ready to merge
